client/consensus/grandpa/interfaces.go

```diff
@@ -42,3 +42,19 @@ type AuxStore interface {
 	// Get Query auxiliary data from key-Value store.
 	Get(key []byte) (*[]byte, error)
 }
+
+type HashI interface {
+	constraints.Ordered
+	IsEmpty() bool
+}
+
+type HeaderI[H HashI, N constraints.Unsigned] interface {
+	ParentHash() H
+	Hash() H
+	Number() N
+}
+
+type HeaderBackend[H HashI, N constraints.Unsigned, Header HeaderI[H, N]] interface {
```

Comment:
why do we need to be generic over the header? why not use the concrete `types.Header` struct?

Reply:
This is because we call the function 
`commitValidationResult, err := finalityGrandpa.ValidateCommit[H, N, S, ID](j.Commit, voters, ancestryChain)`
which requires ancestryChain (which uses the header) to be generic.

However, if there is a way around this let me know and I will update